File: LeetCode/Study Plan/Practice Problems/Hard/Binary Tree Maximum Path Sum/Cpp/binary_tree_max_path_sum.cc
#include <iostream>

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode():val(0), left(nullptr), right(nullptr) {}
    TreeNode(int val):val(val), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {} 
};

int recur(TreeNode *root, int &res) {
    if (root == nullptr) return 0;
    int left = recur(root->left, res);
    int right = recur(root->right, res);
    int l_sum = std::max(root->val, root->val+left);
    int r_sum = std::max(root->val, root->val+right);
    int total_sum = l_sum + r_sum - root->val;
    res = std::max(res, total_sum);
    return std::max(l_sum, r_sum);
}

int binary_tree_max_path_sum(TreeNode *root) {
    if (root == nullptr) return 0;
    int res = INT_MIN;
    recur(root, res);
    return res;
}

int main() {
    TreeNode *root = new TreeNode(1);
    root->left = new TreeNode(2);
    root->right = new TreeNode(3);
    std::cout<<binary_tree_max_path_sum(root)<<"\n";
}

File: LeetCode/Hard/Binary Tree Maximum Path Sum/Java/binaryTreeMaxPathSum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int res;
    public int maxPathSum(TreeNode root) {
        res = Integer.MIN_VALUE;
        recur(root);
        return res;
    }

    private int recur(TreeNode root) {
        if (root == null) return 0;
        int left = recur(root.left);
        int right = recur(root.right);
        int lSum = Math.max(root.val, left+root.val);
        int rSum = Math.max(root.val, right+root.val);
        int total = lSum + rSum - root.val;
        res = Math.max(res, total);
        return Math.max(lSum, rSum);
    }
}
